### ISSUE TITLE
Linux and OSX Open AL Sound Issues

### ISSUE THREAD

**Issue by gamrguy** [bluemagic123/tModLoader#28]
Explained in detail (as much as he could find, anyway) by wuli1986 here:
[Terraria forum post on tModLoader thread](http://forums.terraria.org/index.php?threads/1-3-tmodloader-a-modding-api.23726/page-508#post-993594)

wuli1986 experiences this problem on Linux, and I on OSX. The issue is that, after playing the game for a while, sounds eventually stop being loaded correctly (or something) and fail to play, leading to many annoyances. This does not happen in the vanilla game, and occurs independently of enabled mods. Apparently, this issue is not new; according to wuli1986, it's been there as far back as 0.7.x at the least.

**Comment by gamrguy** [comment]
Confirmed that this issue still occurs in tModLoader 0.8.1.2.

**Comment by Kf4btg** [comment]
I'm wuli1986 over at the Terraria forums (yes, now you know my not-so-secret identity; it's just because I couldn't be bothered to make a new account over there and just used my steam login, itself an ancient remnant from my highschool days...but I digress). I managed to find some time to look into this again.

__TL;DR: Setting "sources=4096" in "~/.alsoftrc" seems to be a fairly adequate workaround for now.__
I would think that file would work on OS X as well as Linux, but I haven't tested it.

Here is where I was going to put the big long writeup I typed out on how I dug through the code for a lot of interconnected projects and libraries and finally ended up compiling a newer version of OpenAL Soft [from source](https://github.com/kcat/openal-soft) which seemed to solve (or prevent) The Mystery of the Disappearing Sound Effects, but then I realized something: I'm an idiot.

Before everyone clamors in agreement, here's why in this case: Terraria doesn't use the system libopenal; it uses the version it ships with (in the "Terraria/lib[64]" folder[s]). I verified this with "lsof $(pidof Terraria.bin.x86)", and it occurred to me that installing the updated OpenAL was simply a catalyst. I had generated a new config file when I updated--one with "sources=4096" set within. The updated version of OpenAL's GUI configuration utility now allows values larger than 256 to be set for the maximum number of sources, but higher values could _already_ be set in the text config file. I thought I had looked into how to do just that the last time I researched this issue, but, well...see the end of the previous paragraph.

So, having set "sources=4096" in my user-config file for openal (here, that's "~/.alsoftrc"), I ran a couple sessions of tModLoader with no mods enabled. Usually, it takes less than 5 minutes of active play to start dropping sound effects, but both of these sessions went for 15 min or longer with no dropped sounds, and with none of the familiar "WARNING: AL SOURCE WAS NOT AVAILABLE. SKIPPING" error messages in the terminal.

Here's what OpenAL reported at the end of session 1:
```
AL lib: (WW) FreeContext: (0x9432d18) Deleting 77 Source(s)
AL lib: (WW) FreeDevice: (0x94474c0) Deleting 113 Buffer(s)
AL lib: (WW) FreeDevice: (0x94474c0) Deleting 1 Filter(s)
```

77 actually surprised me, obviously being quite a bit better than 256. But here's the second session:
```
AL lib: (WW) FreeContext: (0x976f3d0) Deleting 948 Source(s)
AL lib: (WW) FreeDevice: (0x977ea90) Deleting 124 Buffer(s)
AL lib: (WW) FreeDevice: (0x977ea90) Deleting 1 Filter(s)
```

Which is far more what I was expecting. This leads me to hypothesize that the boosted limit of 4096 is enough to ease pressure on the buffers, preventing errors and allowing occasional cleanup to occur as normal. Perhaps the first result would have been higher if I had quit earlier (or later), and similarly the second could have been lower.

However, since even the first result (both the "Source(s)" and "Buffer(s)" counts) are still quite a bit higher than Vanilla, it seems clear that tModLoader is doing _something_ different. With regards to that, [here's a discussion](https://github.com/flibitijibibo/FNA-MGHistory/issues/116) from the old FNA GitHub that I ran across a while ago when researching the "WARNING..." error message. 

Admittedly the issue is a bit old, so I'm not even sure it's applicable any more. This problem was difficult for me to debug since I'm unable to compile and test any changes I'd make to the tModLoader code. Thus, my "debugging" process involved a lot of clicking through source code repos and muttering "Hmm, that could be something..." to myself.

But if the discussion above were relevant, then the only non-mod-specific changes to Terraria's sound processing seem to be [here](https://github.com/bluemagic123/tModLoader/blob/master/patches/merged/Terraria/Main.cs.patch#L205), where there's some manipulation of the ActiveSoundInstances, and [here](https://github.com/bluemagic123/tModLoader/blob/master/patches/tModLoader/Terraria/Main.cs.patch#L2344), where the volume is set to 0 if there are too many sound errors (though of course that wouldn't account for the error message nor for why only SOME of the sounds go missing).

**Comment by gamrguy** [comment]
I can't seem to find any Open AL configuration file on my Mac. If it's there at all, it certainly isn't named .alsoftrc and certainly isn't located in the user home directory.

**Comment by Kf4btg** [comment]
@gamrguy whoops, I probably should have mentioned that the file likely won't exist and will need to be created for this. I'll edit the TL;DR to do just that.

Glad to hear it works for you, too! It's good to have a workaround for this for now; the sound effects really are a vital part of the game's atmosphere. Not having them really detracts from the enjoyment, IMO.

**Comment by bluemagic123** [comment]
I think I remember people reporting this happened only after once a certain version was released. Does this problem still occur in v0.6?

**Comment by yaomtc** [comment]
This has been happening to me, I'm playing with the latest tModLoader on Linux.

**Comment by yaomtc** [comment]
I've tried the workaround, and it does help. However I do still sometimes notice missing sounds, though it's not as bad as before. Thanks for improving the situation, anyway!

**Comment by Godelvatai** [comment]
So, I attempted the fix but I have a limited knowledge on computers and especially mac. I am experiencing the audio issues and attempted to make the file to fix but i am not very good at things and could use a slightly detailed explanation on how to make the file on Mac.

**Comment by Chicken-Bones** [comment]
Hopefully this will be fixed in 0.9

**Comment by Myknakryu** [comment]
Well... It seems that this problem is no more present. Thanks for fixing it.
